pkg/server/etcd.go

```diff
@@ -0,0 +1,102 @@
+package server
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/rancher/k3s/pkg/etcd"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func setETCDLabelsAndAnnotations(ctx context.Context, config *Config) error {
+	t := time.NewTicker(5 * time.Second)
+	defer t.Stop()
+	for range t.C {
+		controlConfig := &config.ControlConfig
+
+		sc, err := newContext(ctx, controlConfig.Runtime.KubeConfigAdmin)
+		if err != nil {
+			logrus.Infof("Failed to set etcd role label: %v", err)
+			continue
+		}
+
+		if err := stageFiles(ctx, sc, controlConfig); err != nil {
+			logrus.Infof("Failed to set etcd role label: %v", err)
+			continue
+		}
+
+		if err := sc.Start(ctx); err != nil {
+			logrus.Infof("Failed to set etcd role label: %v", err)
+			continue
+		}
+
+		controlConfig.Runtime.Core = sc.Core
+		nodes := sc.Core.Core().V1().Node()
+
+		nodeName := os.Getenv("NODE_NAME")
+		if nodeName == "" {
+			logrus.Info("Failed to set etcd role label: node name not set")
+			continue
+		}
+		node, err := nodes.Get(nodeName, metav1.GetOptions{})
+		if err != nil {
+			logrus.Infof("Failed to set etcd role label: %v", err)
+			continue
+		}
+
+		// remove controlplane label if role label exists
+		var controlRoleLabelExists bool
+		if _, ok := node.Labels[MasterRoleLabelKey]; ok {
+			delete(node.Labels, MasterRoleLabelKey)
+			controlRoleLabelExists = true
+		}
+		if _, ok := node.Labels[ControlPlaneRoleLabelKey]; ok {
+			delete(node.Labels, ControlPlaneRoleLabelKey)
+			controlRoleLabelExists = true
+		}
+
+		if v, ok := node.Labels[ETCDRoleLabelKey]; ok && v == "true" && !controlRoleLabelExists {
+			break
+		}
+		if node.Labels == nil {
+			node.Labels = make(map[string]string)
+		}
```

Comment:
I think this needs to go above `// remove controlplane label if role label exists` otherwise you'll get a crash when checking for controlplane labels when Labels is nil.